pkg/server/secrets-encrypt.go

```diff
@@ -0,0 +1,503 @@
+package server
+
+import (
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/rancher/k3s/pkg/daemons/config"
+	"github.com/rancher/wrangler/pkg/generated/controllers/core"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	apiserverconfigv1 "k8s.io/apiserver/pkg/apis/config/v1"
+)
+
+const (
+	EncryptionStart     string = "start"
+	EncryptionPrepare   string = "prepare"
+	EncryptionRotate    string = "rotate"
+	EncryptionReencrypt string = "reencrypt"
+)
+
+const aescbcKeySize = 32
+
+type EncryptionState struct {
+	Stage        string   `json:"stage"`
+	ActiveKey    string   `json:"activekey"`
+	Enable       int      `json:"enable,omitempty"` // -1: no config file, 0: disabled, 1: enabled
+	HashMatch    bool     `json:"hashmatch,omitempty"`
+	HashError    string   `json:"hasherror,omitempty"`
+	InactiveKeys []string `json:"inactivekeys,omitempty"`
+}
+
+type EncryptionRequest struct {
+	Stage  string `json:"stage,omitempty"`
+	Enable bool   `json:"toggle,omitempty"`
+	Force  bool   `json:"force"`
+}
+
+func getEncryptionRequest(req *http.Request) (EncryptionRequest, error) {
+	b, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		return EncryptionRequest{}, err
+	}
+	result := EncryptionRequest{}
+	err = json.Unmarshal(b, &result)
+	return result, err
+}
+
+func encryptionStatusHandler(server *config.Control) http.Handler {
+	return http.HandlerFunc(func(resp http.ResponseWriter, req *http.Request) {
+		if req.TLS == nil {
+			resp.WriteHeader(http.StatusNotFound)
+			return
+		}
+		status, err := encryptionStatus(server)
+		if err != nil {
+			resp.WriteHeader(http.StatusBadRequest)
+			resp.Write([]byte(err.Error()))
+			return
+		}
+		b, err := json.Marshal(status)
+		if err != nil {
+			resp.WriteHeader(http.StatusBadRequest)
```

Comment:
Inability to marshal something we just created would be a server error, not a bad request, I think
```suggestion
			resp.WriteHeader(http.StatusInternalServerError)
```